### PR TITLE
refactor(embodied): move mapping to scheduler, add decoupled env

### PR DESCRIPTION
**Pull request by guozhen1997**
### Description

This modification refactors the communication interfaces among the env worker, reward worker, and rollout worker, unifying these interfaces. In addition, an env decoupled mode is added, which fully decouples the rollout worker from the env worker, supports arbitrary ratios of env workers to rollout workers, and is compatible with both training and evaluation modes. Furthermore, dynamic batching logic has been added to the rollout worker component.

### Motivation and Context

Currently, within RLinf, the communication interfaces among the env worker, reward worker, and rollout worker are relatively independent but have significant overlap, making them rather redundant and difficult for users to understand. Therefore, a refactoring is being carried out. In addition, the env decoupled mode is primarily designed to ensure that the overall training can continue even if a particular env worker fails.

### How has this been tested?

The current PR has undergone relevant testing, and the test results are as follows:
截屏2026-06-12 17 04 58

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Documentation update (Document-only update)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: rlinf/workers/reward/reward_worker.py [hunk 1]
```diff
@@ -256,88 +276,64 @@ def init_worker(self):
...
 
         if self.enable_offload:
             self.model.to("cpu")
 
-    async def recv_merged_reward_input(
-        self, input_channel: Channel, mode: Literal["train", "eval"] = "train"
-    ) -> tuple[dict[str, Any], int]:
-        """Receive all mapped reward inputs and merge on batch dimension.
+    @Worker.timer("compute_image_rewards")
+    def _compute_image_rewards(self, images: torch.Tensor):
```
Comment (qurakchin) [inline rlinf/workers/reward/reward_worker.py]: what's the difference between this and "compute_image_rewards"?

File: rlinf/workers/reward/reward_worker.py [hunk 2]
```diff
@@ -256,88 +276,64 @@ def init_worker(self):
...
-        if main_images is None:
-            raise ValueError(
-                "Reward input dict missing 'main_images' for batch size inference."
-            )
-        if not isinstance(main_images, (np.ndarray, torch.Tensor)):
-            raise TypeError(f"Unsupported main_images type: {type(main_images)}")
+        with torch.no_grad():
+            outputs = self.model(images)
+            probs = outputs["probabilities"]
+            if self._use_reward_prob:
```
Comment (qurakchin) [inline rlinf/workers/reward/reward_worker.py]: It seems that _use_reward_prob is for print debug info. use a proper attr name

File: rlinf/scheduler/worker/worker.py [hunk 3]
```diff
@@ -840,6 +842,353 @@ def connect_channel(self, channel_name: str):
...
+        env_decoupled_mode: bool = False,
+        send_queue_size: int = 0,
+    ):
+        """Route one payload to a worker group through a channel or direct P2P send.
+
+        When ``enable_p2p`` is True, the same routing plan as the channel path is used, but
+        each shard is sent with :meth:`send` to the destination rank instead of ``channel.put``.
+        In that case ``channel`` may be omitted (pass ``None``).
+
+        When ``env_decoupled_mode`` is True, this function will change the plan.entries to send the data to one of the rollout workers.
```
Comment (qurakchin) [inline rlinf/scheduler/worker/worker.py]: what is the "env_decoupled_mode"?

Comment (FxxxxU) [reply rlinf/scheduler/worker/worker.py]: env_decoupled_mode means to make the environment stateless for rollout and reward, so that the whole process will not be blocked.

File: rlinf/workers/reward/reward_worker.py (hunk 2 above)
Comment (FxxxxU) [reply rlinf/workers/reward/reward_worker.py]: delete.

File: rlinf/workers/reward/reward_worker.py (hunk 1 above)
Comment (FxxxxU) [reply rlinf/workers/reward/reward_worker.py]: Two methods handle different situations.

File: rlinf/utils/metric_utils.py [hunk 4]
```diff
@@ -120,7 +124,7 @@ def compute_rollout_metrics(data_buffer: dict) -> dict:
     loss_mask = data_buffer.get("loss_mask", None)
 
     def reduce_metrics(values: torch.Tensor) -> tuple[float, float, float]:
-        device = Worker.torch_platform.current_device()
+        device = _current_device()
```
Comment (qurakchin) [inline rlinf/utils/metric_utils.py]: why change this file

File: rlinf/utils/utils.py [hunk 5]
```diff
@@ -28,11 +28,12 @@
 from torch.distributed.tensor import DTensor
 from torch.optim import Optimizer
 
-from rlinf.scheduler import Worker
 from rlinf.utils.metric_utils import compute_loss_mask
 
 
 def clear_memory(sync=True):
+    from rlinf.scheduler.worker.worker import Worker
```
Comment (qurakchin) [inline rlinf/utils/utils.py:35]: why delete the general import but add these import

File: rlinf/workers/reward/reward_worker.py [hunk 6]
```diff
@@ -33,15 +33,15 @@
...
-from rlinf.utils.comm_mapping import CommMapper
 from rlinf.utils.distributed import all_reduce_dict
 from rlinf.utils.down_sampling import down_sample_batch
 from rlinf.utils.metric_utils import append_to_dict
-from rlinf.utils.nested_dict_process import cat_list_of_dict_tensor
 from rlinf.utils.placement import (
     HybridComponentPlacement,
 )
-from rlinf.utils.utils import clear_memory
+from rlinf.utils.utils import (
```
Comment (qurakchin) [inline rlinf/workers/reward/reward_worker.py:42]: why change this line

File: rlinf/runners/embodied_runner.py [hunk 7]
```diff
@@ -194,12 +194,15 @@ def evaluate(self):
             input_channel=self.env_channel,
             rollout_channel=self.rollout_channel,
         )
-        rollout_handle: Handle = self.rollout.evaluate(
-            input_channel=self.rollout_channel,
-            output_channel=self.env_channel,
-        )
+        env_decoupled_mode = self.cfg.env.train.get("env_mode", None) == "decoupled"
```
Comment (qurakchin) [inline rlinf/runners/embodied_runner.py]: where to call rollout.evaluate in env_decoupled_mode?

File: rlinf/scheduler/worker/worker.py [hunk 8]
```diff
@@ -31,19 +32,21 @@
 import torch
 from omegaconf import OmegaConf
 
+from rlinf.scheduler.hardware import AcceleratorType, AcceleratorUtil, HardwareInfo
```
Comment (qurakchin) [inline rlinf/scheduler/worker/worker.py]: why change these lines

File: rlinf/workers/reward/reward_worker.py [hunk 9]
```diff
@@ -213,14 +213,34 @@ def __init__(self, cfg: DictConfig):
...
+        self._use_reward_prob = self.cfg.reward.get("use_reward_prob", False)
+
+        # check env mode
+        env_mode = self.cfg.env.train.get("env_mode", None)
+        assert env_mode in ["decoupled", None], f"{env_mode} is not supported"
+        self.env_decoupled_mode = env_mode == "decoupled"
+
+        if self.env_decoupled_mode:
+            # save the run-time imformation in communicate channel for decoupled mode
+            self.batch_router = {
```
Comment (qurakchin) [inline rlinf/workers/reward/reward_worker.py:250]: strange type. some batch_router are list, some are dict[list]

File: rlinf/workers/reward/reward_worker.py (hunk 1 above)
Comment (qurakchin) [reply rlinf/workers/reward/reward_worker.py]: what are the different situations

File: rlinf/workers/rollout/hf/huggingface_worker.py [hunk 10]
```diff
@@ -60,12 +59,8 @@ def __init__(self, cfg: DictConfig):
         self.total_num_eval_envs = cfg.env.eval.total_num_envs
         self.num_pipeline_stages = cfg.rollout.pipeline_stage_num
 
-        self.train_batch_size = (
-            self.total_num_train_envs // self._world_size // self.num_pipeline_stages
-        )
-        self.eval_batch_size = (
-            self.total_num_eval_envs // self._world_size // self.num_pipeline_stages
-        )
+        self.train_batch_size = self.total_num_train_envs // self.num_pipeline_stages
```
Comment (qurakchin) [inline rlinf/workers/rollout/hf/huggingface_worker.py:85]: where is the self._world_size?

File: rlinf/workers/rollout/hf/huggingface_worker.py [hunk 11]
```diff
@@ -397,7 +347,15 @@ async def generate_one_epoch(self, input_channel: Channel, output_channel: Chann
         self.update_dagger_beta()
         for _ in range(self.n_train_chunk_steps):
             for _ in range(self.num_pipeline_stages):
-                env_output = await self.recv_env_output(input_channel)
+                env_output = await self.recv_from(
```
Comment (qurakchin) [inline rlinf/workers/rollout/hf/huggingface_worker.py]: add a recv_env_output to merge these self.recv_from and self.send_to

File: rlinf/utils/comm_mapping.py [hunk 12]
```diff
@@ -24,6 +24,57 @@ def build_channel_key(src_rank: int, dst_rank: int, extra: str) -> str:
         """Build a canonical point-to-point channel key."""
         return f"{src_rank}_{dst_rank}_{extra}"
 
+    @staticmethod
+    def decoupled_get_batch_size(
```
Comment (qurakchin) [inline rlinf/utils/comm_mapping.py:28]: it seems that decoupled_get_batch_size is only used in routing.py. why added here?

File: rlinf/workers/reward/reward_worker.py (hunk 1 above)
Comment (FxxxxU) [reply rlinf/workers/reward/reward_worker.py]: _compute_image_rewards uses the reward calculation method defined by RLinf itself, which is used when an ordinary PyTorch model does not define its own reward calculation method. compute_image_rewards calls the reward calculation method defined by the model itself. The two have inconsistent calling interfaces.

File: rlinf/utils/metric_utils.py (hunk 4 above)
Comment (FxxxxU) [reply rlinf/utils/metric_utils.py]: Previously, it was to solve the worker circular reference problem. After the code merge, everything else was deleted. Fix.

File: rlinf/utils/utils.py (hunk 5 above)
Comment (FxxxxU) [reply rlinf/utils/utils.py:35]: Importing 'worker' from outside causes a circular reference, leading to import issues when running experiments.

File: rlinf/workers/reward/reward_worker.py (hunk 6 above)
Comment (FxxxxU) [reply rlinf/workers/reward/reward_worker.py:42]: Running pre-commit run --all-files leads to problems due to the code fix style.

File: rlinf/scheduler/worker/worker.py (hunk 8 above)
Comment (FxxxxU) [reply rlinf/scheduler/worker/worker.py]: Running pre-commit run --all-files leads to problems due to the code fix style.

File: rlinf/workers/reward/reward_worker.py (hunk 9 above)
Comment (FxxxxU) [reply rlinf/workers/reward/reward_worker.py:250]: All batch_router are  dict[list], among which only env_worker's is {}, left as an extension field and not used, reserved for future functionality.

File: rlinf/workers/rollout/hf/huggingface_worker.py (hunk 10 above)
Comment (FxxxxU) [reply rlinf/workers/rollout/hf/huggingface_worker.py:85]: Currently, due to the unified sending/receiving method (recv_from and send_to), self.train_batch_size indicates how much data needs to be transferred at once within these two functions (the batch size transferred at one time across all ranks). Currently, the logic for computing batch split indices is placed inside these two functions, so we need to calculate the data size that each rank needs to send/receive within them. Therefore, we do not divide by self._world_size.

File: rlinf/utils/comm_mapping.py (hunk 12 above)
Comment (FxxxxU) [reply rlinf/utils/comm_mapping.py:28]: Because decoupled_get_batch_size is used to calculate the batch size of the decoupled rollout worker, and it serves the same role as get_dst_ranks and get_src_ranks, it is placed in the CommMapper class.

File: rlinf/runners/embodied_runner.py (hunk 7 above)
Comment (FxxxxU) [reply rlinf/runners/embodied_runner.py]: In env_decoupled_mode, the rollout worker can handle both training and evaluation logic in a mixed manner. However, enabling this mode introduces a serious bug when the rollout worker's eval is called, causing the rollout worker to be launched twice and resulting in unknown GPU kernel occupation issues. Therefore, when env_decoupled_mode is enabled, evaluation for that rollout is disabled. Corresponding modifications have been made under only_eval, and all other embodied_runner instances inherit this method, so no issues have been found in current testing.

File: rlinf/workers/rollout/hf/huggingface_worker.py (hunk 11 above)
Comment (FxxxxU) [reply rlinf/workers/rollout/hf/huggingface_worker.py]: fix

File: rlinf/runners/embodied_runner.py (hunk 7 above)
Comment (qurakchin) [reply rlinf/runners/embodied_runner.py]: re-call of self.rollout.generate in running in env_decoupled_mode

File: rlinf/scheduler/worker/worker.py (hunk 8 above)
Comment (qurakchin) [reply rlinf/scheduler/worker/worker.py]: why the `from ..cluster import` is correct but `from ..hardware import` is bad

File: rlinf/workers/reward/reward_worker.py (hunk 1 above)
Comment (qurakchin) [reply rlinf/workers/reward/reward_worker.py]: why one call self.model, another call self.model.compute_reward? what's the difference

File: rlinf/runners/embodied_runner.py (hunk 7 above)
Comment (FxxxxU) [reply rlinf/runners/embodied_runner.py]: fix. update to async_embodied_runner

File: rlinf/utils/comm_mapping.py [hunk 13]
```diff
@@ -24,6 +24,57 @@ def build_channel_key(src_rank: int, dst_rank: int, extra: str) -> str:
...
+        batch_size: int,
+        src_world_size: int,
+        dst_world_size: int,
+        queue_size: int = 0,
+    ) -> list[int]:
+        """Compute destination ranks and transfer sizes for one source rank."""
+        # in decoupled mode, the src_world_size and dst_world_size are the world size of the source and destination workers.
+        # the batch_size is the total batch size of the source workers.
+        # the return value is a list of batch sizes for thi rank destination to send batch split size.
```
Comment (thereAreDemonsNearby) [inline rlinf/utils/comm_mapping.py]: Is 'thi' a typo ?

File: rlinf/utils/comm_mapping.py [hunk 14]
```diff
@@ -24,6 +24,57 @@ def build_channel_key(src_rank: int, dst_rank: int, extra: str) -> str:
...
+        # while workers with a smaller quantity will aggregate the sent data for batch computation.
+        # The queue length can be set by queue_size
+        split_size = max(src_world_size, dst_world_size)
+
+        assert batch_size % split_size == 0, (
+            f"batch_size ({batch_size}) must be divisible by split_size ({split_size})."
+        )
+
+        if src_world_size >= dst_world_size:
+            # Execute the following code when src_world_size == 1 and dst_world_size == 1.
```
Comment (thereAreDemonsNearby) [inline rlinf/utils/comm_mapping.py]: Is this branch only executed under **when src_world_size == 1 and dst_world_size == 1** or the comment is wrong?

Comment (FxxxxU) [reply rlinf/utils/comm_mapping.py]: fix.

File: rlinf/utils/comm_mapping.py (hunk 13 above)
Comment (FxxxxU) [reply rlinf/utils/comm_mapping.py]: fix.

File: rlinf/utils/comm_mapping.py (hunk 14 above)
Comment (FxxxxU) [reply rlinf/utils/comm_mapping.py]: delete the comment.

File: rlinf/workers/reward/reward_worker.py [hunk 15]
```diff
@@ -256,89 +276,63 @@ def init_worker(self):
...
-            reward_input, last_run_count = await self.recv_merged_reward_input(
-                input_channel, mode="train"
+            merged_data = await self.recv_from(
+                group_name=self.cfg.env.group_name,
+                channel=input_channel,
+                tag="train_reward_obs",
+                async_op=True,
+                batch_size=self.train_batch_size,
+            ).async_wait()
+            merged_images = merged_data.get("images")
```
Comment (thereAreDemonsNearby) [inline rlinf/workers/reward/reward_worker.py]: Is this code safe if 'images' is not in received data?

File: rlinf/workers/reward/reward_worker.py [hunk 16]
```diff
@@ -256,89 +276,63 @@ def init_worker(self):
...
+            last_run = merged_data.get("last_run", None)
+            last_run_count = int(last_run.sum().item()) if last_run is not None else 0
+            rewards = self.compute_threshold_image_rewards(images=merged_images)
```
Comment (thereAreDemonsNearby) [inline rlinf/workers/reward/reward_worker.py]: why the method for reward computing is substituted? Is this backward compatible? (model.compute_reward vs compute_threshold_image_rewards)

File: rlinf/workers/reward/reward_worker.py (hunk 15 above)
Comment (FxxxxU) [reply rlinf/workers/reward/reward_worker.py]: fix.

Comment (FxxxxU) [reply rlinf/workers/reward/reward_worker.py]: add the assert.

File: rlinf/workers/reward/reward_worker.py (hunk 16 above)
Comment (FxxxxU) [reply rlinf/workers/reward/reward_worker.py]: Unify the use of the compute_reward interface.

File: rlinf/runners/async_embodied_runner.py [hunk 17]
```diff
@@ -132,6 +133,24 @@ def update_rollout_weights(self, no_wait=False):
         actor_handle: Handle = self.actor.sync_model_to_rollout()
         self._pending_rollout_weight_sync = (rollout_handle, actor_handle)
 
+    def evaluate(self):
```
Comment (qurakchin) [inline rlinf/runners/async_embodied_runner.py:136]: add an assert to avoid running env_decoupled_mode in sync version embodied_runner.py

**Review by thereAreDemonsNearby — APPROVED** [review]

**Review by qurakchin — APPROVED** [review]

**Review by FxxxxU — APPROVED** [review]
Good job.

**Review by andylin-hao — APPROVED** [review]
